File: strings/0205-isomorphic-strings/java/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isIsomorphic(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }

        Map<Character, Character> forward = new HashMap<>();
        Map<Character, Character> backward = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            char source = s.charAt(i);
            char target = t.charAt(i);

            if (forward.containsKey(source) && forward.get(source) != target) {
                return false;
            }
            if (backward.containsKey(target) && backward.get(target) != source) {
                return false;
            }

            forward.put(source, target);
            backward.put(target, source);
        }

        return true;
    }
}

File: strings/0205-isomorphic-strings/cpp/Solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) {
            return false;
        }

        unordered_map<char, char> forward;
        unordered_map<char, char> backward;

        for (size_t i = 0; i < s.length(); i++) {
            char source = s[i];
            char target = t[i];

            if (forward.count(source) && forward[source] != target) {
                return false;
            }
            if (backward.count(target) && backward[target] != source) {
                return false;
            }

            forward[source] = target;
            backward[target] = source;
        }

        return true;
    }
};
